### PR TITLE
Clarifying the difference between size and length

### PR DESCRIPTION
**Pull request by pontusmelke**
Using `length` on anything other than a path is deprecated, `size` should be used instead. Using `length` still works but there is now a warning associated with invalid usages of `length`.

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/SizeFunction.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    case _: Path    => throw new CypherTypeException("SIZE cannot be used on paths")
+    case s: String  => s.length()
+    case x          => makeTraversable(x).toSeq.length
+  }
+
+  def rewrite(f: (Expression) => Expression) = f(LengthFunction(inner.rewrite(f)))
+
+  def arguments = Seq(inner)
+
+  def calculateType(symbols: SymbolTable): CypherType = CTInteger
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/SizeFunction.scala]: Remove return type annotation for consistency (in this file). 

(Perhaps we should make a decision about return type annotations in our to-be-decided common scala code style.)

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/SizeFunction.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.neo4j.cypher.internal.compiler.v2_3._
+import org.neo4j.cypher.internal.compiler.v2_3.helpers.CollectionSupport
+import pipes.QueryState
+import symbols._
+import org.neo4j.graphdb.Path
+
+case class SizeFunction(inner: Expression)
+  extends NullInNullOutExpression(inner)
+  with CollectionSupport
+{
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/SizeFunction.scala]: I think we normally start (scala) blocks on the same line as the class, ie one line above this.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/Length.scala [hunk 3]
```diff
@@ -36,4 +40,17 @@ case object Length extends Function with SimpleTypedFunction {
 
   def asCommandExpression(invocation: ast.FunctionInvocation) =
     commandexpressions.LengthFunction(invocation.arguments(0).asCommandExpression)
+
+  override def semanticCheck(ctx: SemanticContext, invocation: FunctionInvocation): SemanticCheck =
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/functions/Length.scala]: Same as above; return type annotation on this `def` but not on others in this file.

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/LengthFunctionTest.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ */
+package org.neo4j.cypher.internal.compiler.v2_3.commands
+
+import org.neo4j.cypher.internal.compiler.v2_3._
+import org.neo4j.cypher.internal.compiler.v2_3.commands.expressions.{LengthFunction, Identifier, PathImpl, SizeFunction}
+import org.neo4j.cypher.internal.compiler.v2_3.pipes.QueryStateHelper
+import org.neo4j.cypher.internal.compiler.v2_3.test_helpers.CypherFunSuite
+import org.neo4j.graphdb.{Node, Relationship}
+
+class LengthFunctionTest extends CypherFunSuite {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/LengthFunctionTest.scala:28]: Insert given-when-then comments in these tests.

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/SizeFunctionTest.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ */
+package org.neo4j.cypher.internal.compiler.v2_3.commands
+
+import org.neo4j.cypher.internal.compiler.v2_3._
+import org.neo4j.cypher.internal.compiler.v2_3.commands.expressions.{PathImpl, Identifier, SizeFunction}
+import org.neo4j.cypher.internal.compiler.v2_3.pipes.QueryStateHelper
+import org.neo4j.cypher.internal.compiler.v2_3.test_helpers.CypherFunSuite
+import org.neo4j.graphdb.{Relationship, Node, Path}
+
+class SizeFunctionTest extends CypherFunSuite {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/SizeFunctionTest.scala:28]: Insert given-when-then comments in these tests. Also, I like an empty line between the first method and the class declaration (like in the above file).

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/NewPlannerTestSupport.scala [hunk 6]
```diff
@@ -168,7 +168,8 @@ trait NewPlannerTestSupport extends CypherTestSupport {
 
     assertResultsAreSame(interpretedResult, compiledResult, queryText, "Diverging results between interpreted and compiled runtime")
     assertResultsAreSame(ruleResult, interpretedResult, queryText, "Diverging results between rule planner and interpreted runtime")
-
+    ruleResult.close()
+    interpretedResult.close()
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/NewPlannerTestSupport.scala:172]: These changes are not really related to the PR, right? Perhaps they really belong in the #4550 PR?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/NotificationAcceptanceTest.scala [hunk 7]
```diff
@@ -36,4 +36,27 @@ class NotificationAcceptanceTest extends ExecutionEngineFunSuite with QueryStati
...
+  test("do warn when using length on a pattern expression") {
+    val result = executeWithAllPlanners("explain match (a) where a.name='Alice' return length((a)-->()-->())")
+
+    result.notifications should equal(List(LengthOnNonPathNotification(InputPosition(45, 1, 46))))
+  }
+
+  test("do not warn when using size on a collection") {
+    val result = executeWithAllPlannersAndRuntimes("explain return size([1, 2, 3])")
+    result.notifications shouldBe empty
+  }
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/NotificationAcceptanceTest.scala:72]: What about length and size of strings? Are there tests for that already? (That they don't warn, I mean.)

File: community/cypher/docs/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/FunctionsTest.scala [hunk 8]
```diff
@@ -261,10 +273,10 @@ class FunctionsTest extends DocumentingTestBase {
       syntax = "FILTER(identifier in collection WHERE predicate)",
       arguments = common_arguments,
       text = "`FILTER` returns all the elements in a collection that comply to a predicate.",
-      queryText = """match (a) where a.name='Eskil' return a.array, filter(x in a.array WHERE length(x) = 3)""",
+      queryText = """match (a) where a.name='Eskil' return a.array, filter(x in a.array WHERE size(x) = 3)""",
       returns = "This returns the property named `array` and a list of values in it, which have the length `3`.",
       assertions = (p) => {
-        val array = p.columnAs[Iterable[_]]("filter(x in a.array WHERE length(x) = 3)").toList.head
+        val array = p.columnAs[Iterable[_]]("filter(x in a.array WHERE size(x) = 3)").toList.head
```
Comment (Mats-SX) [inline community/cypher/docs/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/FunctionsTest.scala:279]: The word "length" is still there, on row 277. Should we replace that one too? (It's misspelled since before, too. It should be "... a list of values in it, which has length 3."

File: community/kernel/src/main/java/org/neo4j/graphdb/impl/notification/NotificationCode.java [hunk 9]
```diff
@@ -39,7 +39,10 @@
                     "product, perhaps by adding a relationship between the different parts or by using OPTIONAL MATCH" ),
     LEGACY_PLANNER( SeverityLevel.WARNING,
                     Status.Statement.DeprecationWarning,
-                    "Using PLANNER for switching between planners has been deprecated, please use CYPHER planner=[rule,cost] instead");
+                    "Using PLANNER for switching between planners has been deprecated, please use CYPHER planner=[rule,cost] instead"),
+    LENGTH_ON_NON_PATH( SeverityLevel.WARNING,
+                    Status.Statement.DeprecationWarning,
+                    "Using 'length' on anything that is not a path has been deprecated, please use 'size' instead");
```
Comment (Mats-SX) [inline community/kernel/src/main/java/org/neo4j/graphdb/impl/notification/NotificationCode.java]: I would prefer "Using 'length' on anything that is not a path is deprecated ..."

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/SizeFunction.scala (hunk 1 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/SizeFunction.scala]: No it is done a separate commit in this PR, we are allowed to fix broken code while working on features.

**Comment by pontusmelke** [comment]
@Mats-SX nice feedback!

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/LengthFunctionTest.scala (hunk 4 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/LengthFunctionTest.scala:28]: I added it but I don't think they add much value for small tests, but we should maybe discuss that when deciding on style

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/NotificationAcceptanceTest.scala (hunk 7 above)
Comment (pontusmelke) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/NotificationAcceptanceTest.scala:72]: added test for that
